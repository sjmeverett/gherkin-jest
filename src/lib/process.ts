import {GenericTransformer} from 'stucumber';
import {readFileSync} from 'fs';

require('source-map-support').install({
  environment: 'node',

  retrieveFile: (path: string) => {
    path = path.trim();

    if (!/\.feature$/.test(path)) {
      return null;
    }

    try {
      return process(readFileSync(path, 'utf8'), path);
    } catch (e) {
      return null;
    }
  }
});

export default function process(source: string, filename: string) {
  const transformer = new GenericTransformer({
    featureFn: 'describe',
    scenarioFn: 'it',
    beforeAllFn: 'beforeAll',
    afterAllFn: 'afterAll'
  });

  try {
    return transformer.transform(filename, source);
  } catch (e) {
    if (e.name === 'SyntaxError' && e.location) {
      throw new Error(`${filename} (${e.location.start.line}, ${e.location.start.column}): ${e.message}`)
    } else {
      throw new Error(`${e.name}: ${e.message}`)
    }
  }
}
