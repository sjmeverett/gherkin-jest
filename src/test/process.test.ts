import process from '../lib/process';

describe('process', () => {
  it('should return the correct JS', () => {
    const js = process(`
    @someAttribute
    Feature: this is a test
      As a tester
      I want to test
      So that I can test

      @some-other-attribute
      Scenario: a scenario
        Given some givens
          And some data
            | Table Head 1 | Table Head 2 |
            | Cell 1 | Cell 2|
        When I do a thing
        Then I will succeed
    `, '');

    expect(js).toBe(`
const {cucumber} = require("gherkin-jest");
const co = require("co");

describe("Feature: " + "this is a test", () => {
  beforeAll(() => cucumber.enterFeature(["someAttribute"]));
  afterAll(() => cucumber.exitFeature(["someAttribute"]));
  it("a scenario", co.wrap(function *() {
    const world = cucumber.createWorld();
    yield cucumber.enterScenario(world, ["someAttribute","some-other-attribute"]);
    yield cucumber.rule(world, "some givens");
    yield cucumber.rule(world, "some data", [["Table Head 1","Table Head 2"],["Cell 1","Cell 2"]]);
    yield cucumber.rule(world, "I do a thing");
    yield cucumber.rule(world, "I will succeed");
    yield cucumber.exitScenario(world, ["someAttribute","some-other-attribute"]);
  }));
});`)
  })
})
