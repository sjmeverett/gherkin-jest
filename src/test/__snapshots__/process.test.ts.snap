// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process should return the correct JS 1`] = `
"const {cucumber} = require(\\"stucumber\\"); const _cucumber = cucumber.clone();describe(\\"Feature: using feature files in jest\\", () => {const feature = {\\"name\\":\\"using feature files in jest\\",\\"annotations\\":[{\\"name\\":\\"someAttribute\\",\\"arguments\\":[]}],\\"meta\\":{}};beforeAll(() => {_cucumber.defineRule(\\"I {op:word} {a:int} and {b:int}\\", (world, params) => Promise.resolve().then(() => _cucumber.rule(world, \\"I have numbers <a> and <b>\\", null, params)).then(() => _cucumber.rule(world, \\"I <op> them\\", null, params)));_cucumber.enterFeature(feature);
      });afterAll(() => _cucumber.exitFeature(feature));it(\\"a simple arithmetic test\\", () => {const world = _cucumber.createWorld();const scenario = {\\"name\\":\\"a simple arithmetic test\\",\\"annotations\\":[{\\"name\\":\\"some-other-attribute\\",\\"arguments\\":[]}],\\"meta\\":{}, \\"feature\\": feature};return _cucumber.enterScenario(world, scenario).then(() => _cucumber.rule(world, \\"I have numbers 3 and 4\\", null)).then(() => _cucumber.rule(world, \\"I add them\\", null)).then(() => _cucumber.rule(world, \\"I get 7\\", null)).then(() => _cucumber.exitScenario(world, scenario));});it(\\"a more complex test\\", () => {const world = _cucumber.createWorld();const scenario = {\\"name\\":\\"a more complex test\\",\\"annotations\\":[],\\"meta\\":{}, \\"feature\\": feature};return _cucumber.enterScenario(world, scenario).then(() => _cucumber.rule(world, \\"I have numbers 6 and 4\\", null)).then(() => _cucumber.rule(world, \\"I subtract them\\", null)).then(() => _cucumber.rule(world, \\"I get 2\\", null)).then(() => _cucumber.exitScenario(world, scenario));});it(\\"a more complex test\\", () => {const world = _cucumber.createWorld();const scenario = {\\"name\\":\\"a more complex test\\",\\"annotations\\":[],\\"meta\\":{}, \\"feature\\": feature};return _cucumber.enterScenario(world, scenario).then(() => _cucumber.rule(world, \\"I have numbers 8 and 2\\", null)).then(() => _cucumber.rule(world, \\"I divide them\\", null)).then(() => _cucumber.rule(world, \\"I get 4\\", null)).then(() => _cucumber.exitScenario(world, scenario));});it(\\"using declared rules\\", () => {const world = _cucumber.createWorld();const scenario = {\\"name\\":\\"using declared rules\\",\\"annotations\\":[],\\"meta\\":{}, \\"feature\\": feature};return _cucumber.enterScenario(world, scenario).then(() => _cucumber.rule(world, \\"I add 3 and 4\\", null)).then(() => _cucumber.rule(world, \\"I get 7\\", null)).then(() => _cucumber.exitScenario(world, scenario));});});
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInRlc3QuanMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUMsNEVBQ1Esa01BS0Qsd0dBQ0YsaUVBREUsYUFFRixrREFGRSxHQUxDO0FBQUEsd0RBVUcseVJBQ0YscURBREUsYUFFSCx5Q0FGRyxhQUdILHNDQUhHLHlEQUtRLGlPQUNWLHFEQURVLGFBRVgsOENBRlcsYUFHWCxzQ0FIVywwUkFDVixxREFEVSxhQUVYLDRDQUZXLGFBR1gsc0NBSFcseURBVVIsbU9BQ0YsNENBREUsYUFFSCxzQ0FGRyx5REF6QkgiLCJmaWxlIjoidGVzdC5qcyJ9"
`;
